src/test/java/uk/gov/cshr/controller/account/reactivation/ReactivationControllerTest.java

```diff
@@ -0,0 +1,118 @@
+import uk.gov.cshr.service.ReactivationService;
+import uk.gov.cshr.service.security.IdentityService;
+import uk.gov.cshr.utils.ApplicationConstants;
+import uk.gov.cshr.utils.MockMVCFilterOverrider;
+
+import static org.mockito.Mockito.doNothing;
+import static org.mockito.Mockito.when;
+import static org.powermock.api.mockito.PowerMockito.doThrow;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;
+
+@SpringBootTest
+@AutoConfigureMockMvc
+@RunWith(SpringRunner.class)
+@WithMockUser(username = "user")
+public class ReactivationControllerTest {
+
+    private static final String CODE = "abc123";
+    private static final String EMAIL_ADDRESS = "test@example.com";
+    private static final String DOMAIN = "example.com";
+
+    @Autowired
+    private MockMvc mockMvc;
+
+    @MockBean
+    private ReactivationService reactivationService;
+
+    @MockBean
+    private IdentityService identityService;
+
+    @MockBean
+    private AgencyTokenService agencyTokenService;
+
+    @Before
+    public void overridePatternMappingFilterProxyFilter() throws IllegalAccessException {
+        MockMVCFilterOverrider.overrideFilterOf(mockMvc, "PatternMappingFilterProxy");
+    }
+
+    @Test
+    public void shouldRedirectIfAccountIsAgency() throws Exception {
+        Reactivation reactivation = new Reactivation();
+        reactivation.setEmail(EMAIL_ADDRESS);
+
+        when(reactivationService.getReactivationByCodeAndStatus(CODE, ReactivationStatus.PENDING)).thenReturn(reactivation);
+        when(identityService.getDomainFromEmailAddress(EMAIL_ADDRESS)).thenReturn(DOMAIN);
+        when(identityService.isWhitelistedDomain(DOMAIN)).thenReturn(false);
+        when(agencyTokenService.isDomainInAgencyToken(DOMAIN)).thenReturn(true);
+
+        doNothing().when(reactivationService).reactivateIdentity(reactivation);
+
+        mockMvc.perform(
+                get("/account/reactivate/" + CODE))
+                .andExpect(status().is3xxRedirection())
+                .andExpect(redirectedUrl("/account/verify/agency/" + CODE));
+    }
+
+    @Test
+    public void shouldReactivateAccount() throws Exception {
+        Reactivation reactivation = new Reactivation();
+        reactivation.setEmail(EMAIL_ADDRESS);
+
+        when(reactivationService.getReactivationByCodeAndStatus(CODE, ReactivationStatus.PENDING)).thenReturn(reactivation);
+        when(identityService.getDomainFromEmailAddress(EMAIL_ADDRESS)).thenReturn(DOMAIN);
+        when(identityService.isWhitelistedDomain(DOMAIN)).thenReturn(true);
+
+        doNothing().when(reactivationService).reactivateIdentity(reactivation);
+
+        mockMvc.perform(
+                get("/account/reactivate/" + CODE))
+                .andExpect(status().is3xxRedirection())
+                .andExpect(redirectedUrl("/account/reactivate/updated"));
+    }
+
```

Comment:
There should be a test method for the invited user scenario.

Reply:
This will be covered in the `shouldReactivateAccountIfNotAgency()` test method. We don't to explicitly determine whether they are invited or whitelisted here. 

If the domain is _not_ whitelisted but _is_ agency then we send the user to the verify token screen. 

Otherwise, the user must be whitelisted _or_ invited (but we don't care which specifically) and can reactive their account.